src/main/java/com/amazonaws/encryptionsdk/kms/StandardAwsKmsClientSuppliers.java

```diff
@@ -0,0 +1,252 @@
+ *
+ * or in the "license" file accompanying this file. This file is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the License for the
+ * specific language governing permissions and limitations under the License.
+ */
+
+package com.amazonaws.encryptionsdk.kms;
+
+import com.amazonaws.ClientConfiguration;
+import com.amazonaws.auth.AWSCredentialsProvider;
+import com.amazonaws.encryptionsdk.exception.UnsupportedRegionException;
+import com.amazonaws.services.kms.AWSKMS;
+import com.amazonaws.services.kms.AWSKMSClientBuilder;
+import com.amazonaws.services.kms.model.AWSKMSException;
+
+import java.lang.reflect.InvocationTargetException;
+import java.lang.reflect.Proxy;
+import java.util.HashMap;
+import java.util.HashSet;
+import java.util.Map;
+import java.util.Set;
+
+import static java.util.Objects.requireNonNull;
+import static org.apache.commons.lang3.Validate.notEmpty;
+
+/**
+ * Factory methods for instantiating the standard {@code AwsKmsClientSupplier}s provided by the AWS Encryption SDK.
+ */
+public class StandardAwsKmsClientSuppliers {
+
+    /**
+     * A builder to construct the default AwsKmsClientSupplier that will create and cache clients
+     * for any region. Credentials and client configuration may be specified if necessary.
+     *
+     * @return The builder
+     */
+    public static DefaultAwsKmsClientSupplierBuilder defaultBuilder() {
+        return new DefaultAwsKmsClientSupplierBuilder(AWSKMSClientBuilder.standard());
+    }
+
+    /**
+     * A builder to construct an AwsKmsClientSupplier that will
+     * only supply clients for a given set of AWS regions.
+     *
+     * @param allowedRegions the AWS regions that the client supplier is allowed to supply clients for
+     * @return The builder
+     */
+    public static AllowRegionsAwsKmsClientSupplierBuilder allowRegionsBuilder(Set<String> allowedRegions) {
+        return new AllowRegionsAwsKmsClientSupplierBuilder(allowedRegions);
+    }
+
+    /**
+     * A builder to construct an AwsKmsClientSupplier that will
+     * supply clients for all AWS regions except the given set of regions.
+     *
+     * @param deniedRegions the AWS regions that the client supplier will not supply clients for
+     * @return The builder
+     */
+    public static DenyRegionsAwsKmsClientSupplierBuilder denyRegionsBuilder(Set<String> deniedRegions) {
+        return new DenyRegionsAwsKmsClientSupplierBuilder(deniedRegions);
+    }
+
+
+    /**
+     * Builder to construct an AwsKmsClientSupplier that will create and cache clients
+     * for any region. CredentialProvider and ClientConfiguration are optional and may
+     * be configured if necessary.
+     */
+    public static class DefaultAwsKmsClientSupplierBuilder {
+
+        private AWSCredentialsProvider credentialsProvider;
+        private ClientConfiguration clientConfiguration;
+        private final Map<String, AWSKMS> clientsCache = new HashMap<>();
```

Comment:
This doesn't look to be thread-safe. Should this be a `ConcurrentHashMap`?

Reply:
Good point, I'll make it ConcurrentHashMap. The RegionalClientSupplier that MKPs use also used a ConcurrentHashMap